File: ros/ros2_ws/src/thermo_py/thermo_py/thermo_monitor.py
import rclpy
import datetime
from datetime import datetime, timedelta
from rclpy.node import Node
from thermo_msgs.msg import Temperature


class ThermoMonitor(Node):

    def __init__(self):
        super().__init__('thermo_monitor')
        self.subscription = self.create_subscription(Temperature,
                                                     'temperature',
                                                     self.listener_callback,
                                                     10)

    def to_datestr(self, stamp):
        sec = stamp.sec
        msec = stamp.nanosec / 1_000_000
        time = datetime.fromtimestamp(sec) + timedelta(milliseconds=msec)
        return "{0:%Y-%m-%d %H:%M:%S}".format(time)

    def listener_callback(self, msg):
        time = self.to_datestr(msg.header.stamp)
        temp = msg.temperature
        self.get_logger().info('[%s] temperature=%.2f' % (time, temp))


def main(args=None):
    rclpy.init(args=args)

    subscriber = ThermoMonitor()
    rclpy.spin(subscriber)

    subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros/tutorial/src/thermo_cpp/src/thermo_monitor.cpp
#include <ctime>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/header.hpp"
#include "thermo_msgs/msg/temperature.hpp"

class ThermoMonitor : public rclcpp::Node
{
public:
    ThermoMonitor()
        : Node("thermo_cpp_monitor")
    {
        auto callback = [this](const thermo_msgs::msg::Temperature::SharedPtr msg) -> void
        {
            auto timestamp = this->format_timestamp(msg->header.stamp);
            RCLCPP_INFO(this->get_logger(), "[%s] temperature=%.2f", timestamp.c_str(), msg->temperature);
        };
        subscription_ = this->create_subscription<thermo_msgs::msg::Temperature>("temperature", rclcpp::QoS(10), callback);
    }

private:
    std::string format_timestamp(const builtin_interfaces::msg::Time &time) const
    {
        char date[64];
        time_t sec = time.sec;
        std::strftime(date, sizeof(date), "%Y-%m-%d %H:%M:%S", std::localtime(&sec));
        return date;
    }

private:
    rclcpp::Subscription<thermo_msgs::msg::Temperature>::SharedPtr subscription_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ThermoMonitor>());
    rclcpp::shutdown();
    return 0;
}
